From c1088f625c88ca85d8f3fb87686b46cdd0b6fb68 Mon Sep 17 00:00:00 2001
From: Guan Jing <guanjing6@huawei.com>
Date: Wed, 22 Mar 2023 19:22:50 +0800
Subject: [PATCH 327/921] [Huawei] sched/fair: fix a bug about
 qos_idle_h_nr_running in cfs_rq

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188595

-------------------------------

After enqueue an online task, the cgroup of the online task
is modified to offline, and the qos_level of the online task
will change simultaneously.It will cause decrement
of qos_idle_h_nr_running by 1 when dequeue (it is not yet
an offline task), resulting in an incorrect value
for qos_idle_h_nr_running.
We have modified the criteria of determining offline tasks,
which requires the task to satisfy both the scheduling class
of SHCED_IDLE and the QoS level of -1. Only then the task
will be considered as an offline task.

Signed-off-by: Guan Jing <guanjing6@huawei.com>
---
 kernel/sched/sched.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/sched.h b/kernel/sched/sched.h
index a24a5cd4691f..68384935de27 100644
--- a/kernel/sched/sched.h
+++ b/kernel/sched/sched.h
@@ -2748,7 +2748,7 @@ static inline int qos_idle_policy(int policy)
 
 static inline int task_has_qos_idle_policy(struct task_struct *p)
 {
-	return qos_idle_policy(task_group(p)->qos_level);
+	return qos_idle_policy(task_group(p)->qos_level) && p->policy == SCHED_IDLE;
 }
 #endif
 
-- 
2.17.1

